### ISSUE TITLE
Ruby-style blocks

### ISSUE THREAD

**Issue by sadmac7000** [ceylon/ceylon-spec#1256]
I'm proposing the following new argument syntax specifically for passing a single argument which is a function:

```ceylon
    value foo = [6,7,8,9,10];
    value bar = foo.select{ |x|
        return (x == 7 || (x % 2) == 0);
    };
```

I'd have proposed this earlier, but I feared named argument lists would conflict with it. I think the presence of the argument list at the beginning of the block makes this unambiguous.

The only question would be blocks which take no arguments. Since we need the argument list, we'd need to pass an empty one. We could use `||` or `|void|` for this (I prefer the later).

**Comment by lucaswerkmeister** [comment]
We already have several syntaxes for named arguments function arguments: (ordered from concise to extensive)
```ceylon
foo.select { (x) => x == 7 || x%2 == 0; };
foo.select { (Integer x) => x == 7 || x%2 == 0; };
foo.select { (Integer x) { return x == 7 || x%2 == 0; } };
foo.select {
    Boolean selecting(Integer x) {
        return x == 7 || x%2 == 0;
    }
}
```
In my opinion, there’s no gap here that the Ruby syntax would close.

This syntax is also irregular in only allowing a single argument; if the function logically takes several parameters, the API designer must have written it with this syntax in mind (i. e. curried it).

It also doesn’t look Ceylonic at all. Ceylon strives to be readable even for people not familiar with the language, and mixing syntax from syntactically completely unrelated languages (Ruby and C) doesn’t help with that. (Coming from C/Java, I’ve always found this part of Ruby quite confusing when I stumbled upon it.)

**Comment by gavinking** [comment]
I'm also not especially a fan of this syntax, simply because it's basically a foreign thing that doesn't exhibit much regularity with the rest of the language.

The natural ceylonic syntax for this would be:


```ceylon
    "hello world".each (x) {
        print(x);
        print(x.integer);
    }
```

Unfortunately, that's not really possible, since to the parser it looks like a positional args invocation followed by a named args invocation.

Well, OK OK, we don't currently support named arguments for indirect invocations. So if we were happy to lock in that limitation _forever_, then the truth is that I could indeed force the parser to interpret the above as single invocation with a function as an argument. But:

- are we really ready to lock in that limitation forever?
- I'm not 100% certain that I could really finesse stuff so that a trailing ; is not required, which would be totally lame.

Of course I can squash these doubts by introducing a keyword, for example:

```ceylon
    "hello world".each for (x) {
        print(x);
        print(x.integer);
    }
```

or:

```ceylon
    "hello world".each given (x) {
        print(x);
        print(x.integer);
    }
```

or:

```ceylon
    "hello world".each function (x) {
        print(x);
        print(x.integer);
    }
```

But I'm really not sure how much expressiveness would be gained by this.

**Comment by sadmac7000** [comment]
Instead of a key word, why not an operator?

```ceylon
    "hello world".each : (x) {
        print(x);
        print(x.integer);
    }
```

This has the advantage of not stating its intent so loudly, which means we could generalize it to just mean "pass a single argument.

```ceylon
    "hello world".repeat : 6;
```

Not super useful, but more general constructs tend to feel nicer.

I will attempt to rewrite some bits of ceylon.parse code using a construct like this here to demonstrate its value.

**Comment by ncorai** [comment]
I hate it when languages propose multiple ways to accomplish the same thing. It makes code harder to parse, and devs end up with their own preferred styles, which can lead to less regular code and to version diffs that are more complicated than needed.

It's not like Ceylon doesn't already offer ways to accomplish the same goal, what would the benefit be of adding Ruby syntax to it?

**Comment by chochos** [comment]
I don't like this ruby syntax. Just a personal opinion.

**Comment by luolong** [comment]
While I definitely do not hate the Ruby syntax, I do not really think it adds much in the way of readability to what we already have.

If anything, it adds more confusion. And that only for the sake of pleasing few stray Ruby programmers, not ready to consider alternative syntactic variations to similar concept...

**Comment by sadmac7000** [comment]
I'd be much more amenable to @gavinking 's suggestions at this point.

And I do still think there's a readability improvement.

The point is to avoid this:
```ceylon

someObject.someMethod( (x) {
   ...
   ...
   ...
   ...
   ...
});  // <<< This is bad. Very Bad.
```

Locality is important to readability. Terminators for large blocks are necessarily non-local, and now on top of this, this terminator has three tokens, two of them terminating scopes that don't have much significance to the code reader. Never mind that having such an extremely non-local terminator to a plain argument list is awkward by itself.

**Comment by quintesse** [comment]
I definitely think a construct like this is really nice:

```ceylon
"hello world".each (x) {
    print(x);
    print(x.integer);
}
```

Would we be able to do the following if the function doesn't take any arguments?

```ceylon
time {
    doSomething();
}
```

but if that's all not doable the one using `function` doesn't look too bad either.

**Comment by lucaswerkmeister** [comment]
> ```ceylon
> time {
>     doSomething();
> }
> ```

That looks just like a named argument invocation with an anonymous argument being assigned the return value of `doSomething()`.

**Comment by gavinking** [comment]
As I have already explained above, that is completely ambiguous.

**Comment by quintesse** [comment]
> As I have already explained above, that is completely ambiguous.

That's not how I read what you said, what I read was:

>  then the truth is that I could indeed force the parser to.... [etc]

Now I indeed don't know if I would want to go that far, but I'm just voicing that I like the syntax.

**Comment by gavinking** [comment]
@quintesse like I said, I _might_ be able to force the interpretation of the first code example, at the cost of us never ever supporting named args for indirect invocations. It's not at all clear to me that that's a positive tradeoff.

But your second code example is totally ambiguous because it looks exactly like a direct named args invocation.

**Comment by luolong** [comment]
Adding a keyword like in your last example would not look entirely ugly:
```ceylon
"hello world".each function (x) {
    print(x);
    print(x.integer);
}
```

**Comment by luolong** [comment]
Although I still do not see why this is so bad that we must introduce special syntactic form just to avoid this...
```ceylon
someObject.someMethod( (x) {
   ...
   ...
   ...
   ...
   ...
});  // <<< This is bad. Very Bad.
```

**Comment by gavinking** [comment]
@luolong yeah, I mean the question is: is `function` at the start _really_ that much better than the parens? I admit that it's a _little_ more satisfying, but worth adding extra syntax to a language which already has a lot of syntax? That's hard to see...

**Comment by gavinking** [comment]
I'm resurrecting this zombie issue because I recently realized that since we already allow this:

```ceylon
strings.each((element) {
    print(element);
});
```

(i.e. a multiline function without a `void` or `function` modifier) then that *already* makes indirect named-argument invocations ambiguous!

i.e. one could interpret `(element)` as a reference to a function in parens, and then the `{` would be the beginning of a named argument invocation.

So I think at this point I can just let go of the vague notion that someday I might want to make named args work for indirect invocations.

So what does that mean for this issue? Well, _I think_ it means that we can declare that

```ceylon
xxx.each (element) {
    print(element);
}
```
is unambiguously passing an anonymous function to the method `each()`, which will make @sadmac7000, @quintesse, and me rather happy.

But before we get too excited, first let me play with the grammar some and just be sure that that's really correct, OK?

**Comment by lucaswerkmeister** [comment]
> So I think at this point I can just let go of the vague notion that someday I might want to make named args work for indirect invocations.

Note that, when the “named arguments invocation” interpretation is enforced, the compiler does support calling grouped expressions with named arguments as long as the reference is not indirect once the parentheses are stripped away:
```ceylon
shared void run() {
    (print) { val = "Hello"; };
    // named args, because a function expression isn’t a legal expression statement
}
shared void run2() => ((print)) { val = "Hello2"; };
// named args, because ((x)) isn’t a legal parameter list
shared void run3() => (print) { val = "Hello3"; };
// error: parsed as function expression, print parameter type can’t be inferred, void function can’t return value
```
Perhaps, to avoid confusion, we should restrict the syntax of named arguments so that they’re only legal after a `BaseExpression`, `QualifiedExpression`, or `ConstructorExpression`.

**Comment by gavinking** [comment]
Right, that's certainly what I'm proposing. The above behavior is a bug, resulting from an ambiguity that I did not notice until a couple of days ago.

**Comment by jvasileff** [comment]
So that would address/fix https://github.com/ceylon/ceylon-spec/issues/1449, https://github.com/ceylon/ceylon-compiler/issues/2387, and https://github.com/ceylon/ceylon-js/issues/659

**Comment by gavinking** [comment]
I have implemented the above syntax on the branch `1256`. On that branch, the following code compiles and runs:

```ceylon
    shared void run() {
        {"hello", "world", "goodbye"}
                .each (element) {
            print(element);
        }
    }
```

**Comment by gavinking** [comment]
However, I actually have an alternate syntax proposal, that I believe would be significantly easier on the parser (and that probably also means a better experience in the IDE). 

How about something like this:

```ceylon
(element in list.each) {
    print(element);
}
```

Second example:

```ceylon
(tx in sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

Third example:

```ceylon
(x, y in collection.eachPair) {
    if (x==y) { ... }
}
```

Where there are various options we can argue over for the token I've written as `in` here.

The idea here is that this is reflective of the syntax we use for other control structures in the language, and it's clear that the *purpose* of this feature is to be able to use a higher-order function as a control structure.

Thoughts?

**Comment by gavinking** [comment]
Other options include stuff like:

```ceylon
(tx: sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

and

```ceylon
(tx <- sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

and

```ceylon
(tx ~ sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

And

```ceylon
(tx := sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

and

```ceylon
(tx for sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

If the chosen token is _not_ already an infix operator in Ceylon, it's a lot better from the PoV of the IDE.

**Comment by gavinking** [comment]
Or even:

```ceylon
(tx; sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

**Comment by gavinking** [comment]
Surprisingly, this doesn't look terrible:

```ceylon
(tx @ sql.transaction) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

```ceylon
(element @ list.each) {
    sql.update(...);
    tx.rollBackOnly = true;
}
```

**Comment by quintesse** [comment]
Personally I find the parentheses a bit distracting, we don't write 

```ceylon
(while x < y) {
}
```

Although like everything we'd probably get used to it.

Can those methods have more than one argument? For example:

```ceylon
(tx in sql.transaction(context)) {
    ...
}
```

**Comment by jvasileff** [comment]
This is a feature that encourages mutability and side effects, and therefore I don't think it should be added. Ceylon already has a perfectly good `for` (and other) statements for use when side effects are needed or desired, and by using these statements, the intention of the contained code is clear.

*Perhaps* it could be used something like `Contextual`, but any improvement in brevity would be offset by a loss in clarity by not using `try`. And, generally, I'd want a `contextual` *function* to return a value.

I would feel completely different on this if the idea were to also support expressions.

**Comment by gavinking** [comment]
Note: if other people feel like John, and _don't_ want this feature in the language, I'm OK with that. It's not something that I personally have to have. I _do_ feel some discomfort when I use `each()` in Ceylon, but not so much that it really bothers me.

**Comment by FroMage** [comment]
I'm not convinced it would be a good thing to add.

**Comment by DiegoCoronel** [comment]
>Note: if other people feel like John, and don't want this feature in the language

-1, this is something that i would not consider predictable, its not hard to read, but i dont like too

**Comment by luolong** [comment]
Let me play with this syntax a little and offer up two alternative syntaxes for the same thing.

Accorging to @gavinking we could have this syntax

```ceylon
shared void run() {
    {"hello", "world", "goodbye"}.each (element) {
        print(element);
    }
}
```

versus this one:
```ceylon
shared void run() {
    (element in {"hello", "world", "goodbye"}.each) {
        print(element);
    }
}
```

Of the two I actually like the first option better...

**Comment by gavinking** [comment]
@luolong well that's what is implemented in the branch. You can try it out if you like (it all works perfectly AFAICT). But note that it seems that there is not really overwhelming support for it.

**Comment by luolong** [comment]
I thing (and I may be wrong), that the latest round of _nays_ was against the second syntactic version, which is confusing and does not really match well with the rest of the language ...

**Comment by davidfestal** [comment]
I'm OK with any syntax that would *not* prevent the ability to use named arguments in multiple parameter lists in the future.

It seems to me that enabling named arguments in multiple parameter lists, would *greatly* enhance the definition of DSLs.
I've been trying to define a Named-argument-based DSL for Camel, and I would have liked this several times.

Named arguments are the only way (afaik) to define DSLs in Ceylon, and I think that we should not limit their future enhancements just to introduce a syntax change as this one. 

And besides DSLs, being able to use a `function` definition in a named argument list for the `accumulating` argument of the `Iterable.fold()()` method would be cool.

**Comment by davidfestal** [comment]
as for a syntax, I would have prefereed with a keyword (I used `with` here) : 
```ceylon
shared void run() {
    {"hello", "world", "goodbye"}
            .each with (element) {
        print(element);
    }
}
```

@gavinking : if we introduce a keyword, do this solve the problem of having to prevent named arguments in MPLs ?

**Comment by luolong** [comment]
So, if I understand you correctly, the issue here is mixing the named argument syntax form with "ruby style" closure syntax.

So we do actually have two cases here -- either the argument list has multiple arguments of any type or it has a single callable argument.

So for example, lets take `loop` function from language module for example:

```ceylon
// the signature of loop function:
shared {Element+} loop<Element>(Element first)(Element|Finished next(Element element))

// The "old way of calling"
value naturals = loop(1)((number) { number + 1 } );

// We can call this instead:
```

**Comment by gavinking** [comment]
@davidfestal sure, with a new keyword all problems disappear.

**Comment by jvasileff** [comment]
@luolong my objection is not specific to any of the proposed syntaxes.

**Comment by davidfestal** [comment]
@luolong : my issue was that in a more general way, allowing the proposed syntax in the grammar would make the use of named arguments in MPLs impossible in the future. And I find this would a bad thing.

**Comment by jvasileff** [comment]
This discussion has drifted from @sadmac7000's original proposal, which I support and would like to comment on. The proposal would allow some form of:

```ceylon
print {
    numbers.select (x) {
        return x == 7 || (x % 2) == 0;
    };
};
```
and
```ceylon
[Integer*] computeSquares({Integer*} ints)
    =>  ints.collect (i) {
            return i*i;
        };
```
and similar which I've found to be fairly common needs.

Now, a desire to eliminate the trailing `;` seems to be an issue WRT to supporting expressions. But, lines ending in `};` are *very* common in Ceylon, and are here to stay. There are 613 occurrences in the SDK:
```
$ git grep '^\s*};$'  -- *.ceylon  | wc -l
     613
```
and over 900 in my dart backend code. In fact, they are so common, that the example with `each` in https://github.com/ceylon/ceylon-spec/issues/1256#issuecomment-78562377 looks bare to me: a method call that does not end in a `;`. If a trailing `;` is good enough for named argument invocations, it should be good enough here.

What I've found *does* cause problems in practice (extra time working through syntax errors, etc.) are lines ending in `});`, which I think @sadmac7000's proposal aims to eliminate. In the SDK, there are 23 occurrences of `});`, of which about 10 would be affected by this. Although I suspect block-stye anonymous functions would be more common if the parenthesis were not required.

**Comment by gavinking** [comment]
I kinda like @jvasileff's proposal and it's easy to implement. The resulting code is very readable. The problem is that we would now have two reasonable ways to write those code examples:

    value squares = ints.filter (i) { return i>0; }.collect (i) { return i*i; };

And:

    value squares = ints.filter((i) => i>0).collect((i) => i*i);

And they're _really_ dissimilar to each other:

> So, wait, I need the parens when I use the "short" form with a fat arrow and I _don't_ need them when I have a block.

That's a bit ... strange.

**Comment by fwgreen** [comment]
The ability to have an API like...
```Ceylon
get("/:name") (request, response) {
        if (exists name = request.variables["name"],
            exists todo = todos[name]) {
            return todo;
        } else {
            return halt(404);
        }
};
```
instead of...
```Ceylon
get("/:name", (request, response) {
        if (exists name = request.variables["name"],
            exists todo = todos[name]) {
            return todo;
        } else {
            return halt(404);
        }
});
```
would be very welcome.  Assuming this proposal would make it possible?

**Comment by gavinking** [comment]
@fwgreen well, yes, that would be possible. Assuming that `get()` has the following signature:

```ceylon
    void get(String path)(Result (Request,Response) handler) { ... }
```

**Comment by jvasileff** [comment]
I think most would find it reasonable that argument lists must have at least one final closing delimiter. In this case, it's a `}` instead of `)`, just as it is with named argument lists.

**Comment by ncorai** [comment]
My problem with
`value squares = ints.filter (i) { return i>0; }.collect (i) { return i*i; };`
is that, even if it's not supported, it still _looks like_ positional argument invocations followed by named argument invocations. If I have to remember the context to be able to parse the syntax, then the language has failed both me and its tag line.

If what you're trying to achieve cannot fit in a single expression that could be used as in
`value squares = ints.filter((i) => i>0).collect((i) => i*i);`
(which I assume is the purpose of the alternate syntax), then you should probably start by extracting that logic.

**Comment by luolong** [comment]
I see little or no ambiguity the curly braces syntax in function invocation can be only one of two things. 

Either it is the named argument invocation or it's a inline closure for a single Callable argument.

Latter would also nicely fit in as a tool for DSL designers for introducing new block statement like expressions. 

For example what about this

    shared T iff (T(Boolean condition) closure) given T satisfies Anything {...};
